modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/A11y.es.js

```diff
@@ -0,0 +1,272 @@
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {ClayButtonWithIcon} from '@clayui/button';
+import ClayIcon from '@clayui/icon';
+import ClayPopover from '@clayui/popover';
+import observeRect from '@reach/observe-rect';
+import axe from 'axe-core';
+import React, {useEffect, useMemo, useRef, useState} from 'react';
+import ReactDOM from 'react-dom';
+
+import './A11y.css';
+
+function segmentAlertsByNode(alerts) {
+	const nodes = alerts.flatMap((alert) =>
+		alert.nodes.flatMap((node) => node.target)
+	);
+
+	return nodes.map((node) => {
+		const alertsByNode = alerts.filter((alert) =>
+			alert.nodes.some((n) => n.target.includes(node))
+		);
+
+		return {
+			alerts: alertsByNode,
+			node,
+		};
+	});
+}
+
+const Overlay = React.forwardRef(({style, ...othersProps}, ref) =>
+	ReactDOM.createPortal(
+		<div
+			{...othersProps}
+			className="a11y-overlay"
+			ref={ref}
+			style={style}
+		/>,
+		document.body
+	)
+);
+
+function Alert({alerts, target}) {
+	const [visible, setVisible] = useState(false);
+	const [bounds, setBounds] = useState();
+
+	useEffect(() => {
+		const node = document.querySelector(target);
+
+		if (!node) {
+			return;
+		}
+
+		const setOverlayBounds = (bounds) => {
+			if (bounds.x > 0 || bounds.y > 0) {
+				const borderRadius = window
+					.getComputedStyle(node, null)
+					.getPropertyValue('border-radius');
+
+				setBounds({
+					'border-radius': borderRadius,
+					height: bounds.height,
+					left: bounds.left,
+					top: bounds.top,
+					width: bounds.width,
+				});
+			}
+		};
+
+		const {observe, unobserve} = observeRect(node, (targetRect) => {
+			setOverlayBounds(targetRect);
+		});
+
+		const bounds = node.getBoundingClientRect();
+		setOverlayBounds(bounds);
+
+		observe();
+
+		return () => {
+			unobserve();
+		};
+	}, [target, alerts]);
+
+	return (
+		<ClayPopover
+			className="popover-a11y"
+			header={
+				<div className="autofit-row autofit-row-center">
+					<div className="autofit-col autofit-col-expand">
+						<span>Accessibility violation</span>
+					</div>
+					<div className="autofit-col">
+						<ClayButtonWithIcon
+							borderless
+							displayType="secondary"
+							onClick={() => setVisible(!visible)}
+							small
+							symbol="times"
+						/>
+					</div>
+				</div>
+			}
+			onShowChange={setVisible}
+			show={visible}
+			trigger={<Overlay style={bounds} />}
+		>
+			{alerts.map((alert) => {
+				const [{all, any}] = alert.nodes.filter((node) =>
+					node.target.includes(target)
+				);
```

Comment:
More nested iteration here... `map` + `filter` + `includes` + this entire component being rendered in a loop via another `map` — how snappy does this feel in practice?

Reply:
I didn’t even check if this is having a very big problem, the perceptions of slowness come more when the `axe-core` is really analyzing, as I didn’t get pages with a large amount of violations at most 5, this ends up not impact on performance.

Reply:
But anyway I will try to improve this and see if I can get a page with many violations.